src/chocolatey/infrastructure/services/XmlService.cs

```diff
@@ -65,30 +98,45 @@ public void serialize<XmlType>(XmlType xmlType, string xmlFilePath, bool isSilen
                 () =>
                 {
                     var xmlSerializer = new XmlSerializer(typeof(XmlType));
-                    //var textWriter = new StreamWriter(xmlUpdateFilePath, append: false, encoding: new UTF8Encoding(encoderShouldEmitUTF8Identifier: false))
-                    var textWriter = new StreamWriter(xmlUpdateFilePath, append: false, encoding: new UTF8Encoding(encoderShouldEmitUTF8Identifier: true))
-                    {
-                        AutoFlush = true
-                    };
 
-                    xmlSerializer.Serialize(textWriter, xmlType);
-                    textWriter.Flush();
+                    // Write the updated file to memory
+                    using(var memoryStream = new MemoryStream())                        
+                    using(var textWriter = new StreamWriter(memoryStream, encoding: new UTF8Encoding(encoderShouldEmitUTF8Identifier: true)))
+                    {
+                        xmlSerializer.Serialize(textWriter, xmlType);
+                        textWriter.Flush();
 
-                    textWriter.Close();
-                    textWriter.Dispose();
+                        memoryStream.Position = 0;
+                        
+                        // Grab the hash of both files and compare them.
+                        var originalFileHash = _hashProvider.hash_file(xmlFilePath);
+                        if (!originalFileHash.is_equal_to(_hashProvider.hash_stream(memoryStream)))
+                        {
+                            // If there wasn't a file there in the first place, just write the new one out directly.
+                            if(string.IsNullOrEmpty(originalFileHash))
+                            {
+                                using(var updateFileStream = _fileSystem.create_file(xmlFilePath))
+                                {
+                                    memoryStream.Position = 0;
+                                    memoryStream.CopyTo(updateFileStream);
+                                    return;
+                                }
+                            }
 
-                    if (!_hashProvider.hash_file(xmlFilePath).is_equal_to(_hashProvider.hash_file(xmlUpdateFilePath)))
-                    {
-                        _fileSystem.copy_file(xmlUpdateFilePath, xmlFilePath, overwriteExisting: true);
+                            // Otherwise, create an update file, and resiliently move it into place.
+                            var tempUpdateFile = xmlFilePath + ".update";
+                            using(var updateFileStream = _fileSystem.create_file(tempUpdateFile))
+                            {
+                                memoryStream.Position = 0;
+                                memoryStream.CopyTo(updateFileStream);
+                            }
+                            _fileSystem.replace_file(tempUpdateFile, xmlFilePath, xmlFilePath + ".backup");
+                        }

```

Comment:
I mostly keep it because it allows us to recover if a future operation fails for some reason.